### PR TITLE
feat(UI): confetti when project complete

### PR DESCRIPTION
**Pull request by a2937**
Checklist:

- [X] I have read and followed the [contribution guidelines](https://contribute.freecodecamp.org).
- [X] I have read and followed the [how to open a pull request guide](https://contribute.freecodecamp.org/#/how-to-open-a-pull-request).
- [X] My pull request targets the `main` branch of freeCodeCamp.
- [X] I have tested these changes either locally on my machine, or GitPod.

Related to #46178

### CODE REVIEW

**Review by naomi-lgbt — CHANGES_REQUESTED** [review]
Okie this is super cool and I love confetti a lot.

But because we're hooking this into the completion modal, it might be showing when we don't want it to. My understanding is we want it to show when a project (or block) is completed - but maybe I am misunderstanding and we only want it on certification projects. 

That functionality *works* on this PR, if you're doing the Responsive Web Design challenges. However, switch to the JavaScript challenges and you see confetti after each challenge, because the completion modal shows for every challenge that isn't a multi-file challenge.

**Comment by a2937** [comment]
> Okie this is super cool and I love confetti a lot.
> 
> But because we're hooking this into the completion modal, it might be showing when we don't want it to. My understanding is we want it to show when a project (or block) is completed - but maybe I am misunderstanding and we only want it on certification projects.
> 
> That functionality _works_ on this PR, if you're doing the Responsive Web Design challenges. However, switch to the JavaScript challenges and you see confetti after each challenge, because the completion modal shows for every challenge that isn't a multi-file challenge.


Okay but what should I hook into instead so it only displays on certification projects?

**Comment by naomi-lgbt** [comment]
We could potentially leave it in the completion modal - I believe we'd want to do something similar to what the progress bar component does.

That is, we can [check if the challenge is a certification project](https://github.com/freeCodeCamp/freeCodeCamp/blob/175f039fd6bd82857b5d006de338dacfd419be9d/client/src/components/ProgressBar/progress-bar.tsx#L58) and use that to conditionally display the confetti.

The completion modal receives the dashed name from the challenge, which *could* be used for that logic. But it might be better to update the component to take an `id` prop for this purpose.

**Comment by a2937** [comment]
Hello Naomi. I have a question. I'm not quite familiar on how the id was mapped to the progress bar. Can you explain how it was done so I can do something similar in the completion model?

**Comment by naomi-lgbt** [comment]
It looks like [it is pulled from the state](https://github.com/freeCodeCamp/freeCodeCamp/blob/46d1f1122e38b30e467670b361097a65d71a0f05/client/src/components/ProgressBar/progress-bar.tsx#L15)

**Comment by a2937** [comment]
Will there be any complications if I leave it in the completion model? It made the most sense at the time of the implementation to put it in there. 

Also, regarding my implementation,  how does it look? Anything that needs to be improved?

**Comment by naomi-lgbt** [comment]
The implementation looks good. Looks like this needs a merge to resolve the conflicts.

Also, I think you're running `pnpm` version `8.6.0-8.6.2` - you'll want to update with `pnpm i -g pnpm` to get a version that includes the fix to the lockfile versioning.

**Comment by ahmaxed** [comment]
It can stay where it is. A cypress test could be added to lock the feature.

**Comment by a2937** [comment]
Okay so all I need is tests to ensure that 
A) The confetti won't play after a regular Javascript challenge 
B) The confetti will play after completing a project? 

Any other scenarios I need to check?

**Comment by ahmaxed** [comment]
Those should be enough.

File: client/src/templates/Challenges/components/completion-modal.tsx [hunk 1]
```diff
@@ -155,8 +163,10 @@ class CompletionModal extends Component<
 
     if (isOpen) {
       executeGA({ event: 'pageview', pagePath: '/completion-modal' });
+      if (isCertificationProject(id)) {
```
Comment (ahmaxed) [inline client/src/templates/Challenges/components/completion-modal.tsx:166]: I think you would also need to check if the id is not in completed challenges, so it fires only when submitted for the first time.

Comment (a2937) [reply client/src/templates/Challenges/components/completion-modal.tsx:166]: That would be wise.

**Comment by a2937** [comment]
How do I get this back up to date with the main branch?

**Comment by naomi-lgbt** [comment]
Through the command line:

```bash
git checkout confetti
git fetch upstream
git merge upstream/main
```

You'll need to resolve the conflicts in the client/package.json and the completion-modal.tsx file. Don't try to resolve the conflicts in the pnpm-lock file it gets way too messy - instead, after resolving the package.json, delete the pnpm-lock file and run `pnpm install` to get a fresh one in line with the package.json files. 

If you don't have the `upstream` configured, you can `git remote add upstream <url>`, using freeCodeCamp's repo URL (SSH if configured, HTTPS if not).

**Comment by a2937** [comment]
I can't merge.The hook is getting in the way

**Comment by a2937** [comment]
Okay I think I've got it nice and up to date now.

File: client/src/templates/Challenges/components/completion-modal.tsx [hunk 2]
```diff
@@ -78,6 +81,11 @@ interface CompletionModalState {
   downloadURL: null | string;
 }
 
+const isCertificationProject = (id: string) =>
+  certMapWithoutFullStack.some(cert => {
+    return cert.projects.some((project: { id: string }) => project.id === id);
+  });
```
Comment (naomi-lgbt) [inline client/src/templates/Challenges/components/completion-modal.tsx:87]:
```suggestion
const isCertificationProject = (id: string) =>
  Object.values(certsToProjects).some(cert =>
    cert.some(project => project.id === id)
  );
```

File: client/src/templates/Challenges/components/completion-modal.tsx [hunk 3]
```diff
@@ -28,6 +28,8 @@ import ProgressBar from '../../../components/ProgressBar';
 import GreenPass from '../../../assets/icons/green-pass';
 
 import './completion-modal.css';
+import { fireConfetti } from '../../../utils/fire-confetti';
+import { certMapWithoutFullStack } from '../../../resources/cert-and-project-map';
```
Comment (naomi-lgbt) [inline client/src/templates/Challenges/components/completion-modal.tsx:32]:
```suggestion
import { certsToProjects } from '../../../../config/cert-and-project-map';
```

**Review by naomi-lgbt — CHANGES_REQUESTED** [review]
The config setup was changed - this *should* fix the issues, I think.

**Review by naomi-lgbt — APPROVED** [review]
I'm happy with this. If Ahmad is good, I'll resolve the lockfile and merge.

**Comment by ahmaxed** [comment]
There are currently two confettis pop: one for completing the project and one for submitting it.
Having only one for completing the project should be sufficient.

**Comment by a2937** [comment]
I think I fixed that yesterday.

**Review by ahmaxed — APPROVED** [review]

**Comment by ahmaxed** [comment]
@naomi-lgbt LGMT
@a2937, thank you for your hard work here.
If you want to improve on this pr so more people can enjoy your confetti, I suggest doing the followings:
- Move the logic to the related saga where modal popping logic exist
- change the logic so the confetti would show up at the end of each block in addition to project completion
- tag me so I could add an AB test to measure this features performance.

**Comment by a2937** [comment]
I'd like it to get this merged as soon as possible then a new issue with potential improvements listed.

**Comment by a2937** [comment]
@ahmaxed  Is this acceptable?

**Comment by a2937** [comment]
I can't figure out where to move my logic too. Sorry. :/

**Comment by ahmaxed** [comment]
Apologies for the confusion. This PR is good to go as soon as all the tests are passed. 
The new suggestions are to be implemented in upcoming iterations.
Once this PR is merged, I will create an issue with further instructions and tag you there. 
I will revert the changes from the last two commits accordingly.

File: client/src/components/ProgressBar/progress-bar.tsx [hunk 4]
```diff
@@ -70,6 +71,12 @@ function ProgressBar({
       : t('learn.percent-complete', {
           percent: completedPercent
         });
+  if (
+    completedPercent == 100 &&
+    completedChallengesInBlock == totalChallengesInBlock
+  ) {
+    fireConfetti();
+  }
```
Comment (ahmaxed) [inline client/src/components/ProgressBar/progress-bar.tsx:79]:
```suggestion
```

File: client/src/components/ProgressBar/progress-bar.tsx [hunk 5]
```diff
@@ -10,6 +10,7 @@ import {
   completedPercentageSelector
 } from '../../templates/Challenges/redux/selectors';
 import { liveCerts } from '../../../config/cert-and-project-map';
+import { fireConfetti } from '../../utils/fire-confetti';
```
Comment (ahmaxed) [inline client/src/components/ProgressBar/progress-bar.tsx:13]:
```suggestion
```

**Comment by a2937** [comment]
Thank you. Can you help get this up to date again then merge before anything else silly happens? :/

**Comment by ahmaxed** [comment]
@a2937, the related tests are not passing. Could you please investigate them?

**Comment by a2937** [comment]
Sure thing. I'd be happy to look into the issue. It seems to be working fine in the browser, so perhaps there might be an issue in the test itself.

**Comment by shootermv** [comment]
im not experienced anough, but i wish to know if i could help somehow ...

**Comment by a2937** [comment]
> im not experienced anough, but i wish to know if i could help somehow ...

If you look at the current test results, I fixed the multi-file cert project tests last night meaning it should be safe to merge now.

**Comment by a2937** [comment]
@ahmaxed

**Comment by Sboonny** [comment]
There is a misunderstanding

> If you look at the current test results

When shootermv posted his comment, the container test was failing, which needed to be restarted.

**Comment by a2937** [comment]
I see. Thank you for the clarification. My apologies to shootermv.  I appreciate the offer to help and apologize for being a little snippy.

**Comment by moT01** [comment]
just saw the confetti for the first time - looks awesome 🎉

**Pull request merged**